docs/blog/2021-09-28-building-a-startup-with-rust.md

```diff
@@ -0,0 +1,163 @@
+author_image_url: https://avatars.githubusercontent.com/u/14791384?s=460&v=4
+tags: [rust, startup]
+description: This blog post is a compilation of thoughts around building a company with Rust
+image: https://storage.googleapis.com/getsynth-public/media/api.jpg
+hide_table_of_contents: false
+---
+
+When building a company you are setting out to fundamentally solve a problem.
+For this reason, engineers have been systematically attracted by this romantic
+idea of changing the world with your brain and a laptop. We are at heart problem
+solvers.
+
+As engineers we can (and most of us have) become zealous at times about our
+solutions to these problems. We have pragmatists who just get stuff done - they
+address the symptom fast and effectively. We have idealists who will grind at an
+elegant scalable solution and try to treat the disease. Whichever camp you
+subscribe to, at a certain point you need to form an opinion about which
+technologies you are going to use to solve the problems you see in the world -
+and this opinion will inevitably cause contention.
+
+Conventional wisdom is to 'use the right tool for the job'. The choice of
+programming language for example, depends on the domain of the problem you are
+trying to solve. If you're implementing some algorithm, in a secluded project,
+it's easy to make the case about what the language for the job may be. You can
+run a benchmark and literally test the execution time for each candidate
+language (if you're optimising for execution time). You can persuade yourself
+you've made a rational and 'objectively correct' decision.
+
+However, in the context of building a business, your optimisation function is a
+high-dimensional mess involving performance, development velocity, hiring,
+server costs, ecosystem, tooling, support, licenses etc. You can assign weights
+to what is most important for your business, but at the end of the day the
+decision is inevitably qualitative.
+
+At Synth, we're working on building the best data generator in the world. We
+made a conscious decision to use Rust for our main line of products. After more
+than a year of building I've had the opportunity to see Rust at its best and
+worst in the context of starting a company - this post is a compilation of
+these (at times cynical) thoughts.
+
+## Development Velocity
+
+Rust has a *really* steep learning curve. Coming from an OO background it took
+me *months* to become productive in Rust. This was incredibly frustrating for me
+as I felt that my lack of productivity was impacting the team, which it was.
+Even when you eventually do become productive (and you will), Rust forces you to
+really think deeply about what you're doing and things inevitably take longer to
+get over the line. A poorly thought out design decision today can come back to
+haunt you months later. What should be a simple change or refactor can end up
+resulting in complete tear down as you try to appease the borrow checker. This
+is deadly.
```

Comment:
On the other hand, refactors in Rust can often be pretty painless with a bit of planning, because one can change the types so the compiler will direct the following restructuring of operations.

I also don't feel that I need to appease borrowck that much, but that's perhaps my unconscious internal borrowck doing the work before I even start typing.